Parse `fHIDE` declarations

AntonJohansson opened the pull request:
Lexes and parses type declarations previously hidden in the `fHIDE` macro, such as `fHIDE(int tmp;)`. This type information is used to enforce and generate the correct `bit_width` and `signedness` for all variables. Undeclared variables are no longer allowed and result in errors.

Two notable exceptions are the immediate value `i` used in for-loops, and the effective address `EA`, both of which are assumed to always be 32-bit and signed.

Commit:
Redeclared fHIDE macro and added lexing/parsing

The new expressions that were previously hidden in fHIDE are now lexed
and parsed. This primarily includes variable declarations of integer
type such as `int j;`, but also some `if` statements and calls to
a `C`-function `fatal(...)`. This function is currently NOP'd.

Nothing is done with the additional type information yet. Declarations
and assigments such as `int j = 0` are currently treated as `j = 0`
where the type of `j` is the default signed 32-bit int. This results in
two failing tests that use `fSETHALF(n, tmp, ...)` where `tmp = 0` is
now included which sets `tmp` as 32-bit and not 64-bit as `fSETHALF`
would initialize it.  This is easily fixed in the future by including
the new type information.

Signed-off-by: Anton Johansson <anjo@rev.ng>

Commit:
Now uses the parsed type info to set sign/width

Allocates a new varid of specified sign/width when a variable is
declared.

Now also enforcing signedness in `gen_bin_op` for binary expressions
with variables. This probably needs to be done in more places.

Other minor changes includes setting default signedness and width of
VAR_ID tokens to `signedness = UNKNOWN_SIGNEDNESS` and `bit_width = 0`
to ensure we set all types via declarations. The sign of the immediate
`i` was also set to `SIGNED`.

Signed-off-by: Anton Johansson <anjo@rev.ng>

Commit:
Added cast to `fASHIFTR` macro

Needed because we're now taking types of variables into account and
this resulted in the incorrect semantics of the `asr_i_p_rnd`
instruction. This because some variables were declared unsigned and
expected to be casted in the `fASHIFTR` macro, which we did not do.

Signed-off-by: Anton Johansson <anjo@rev.ng>

Commit:
Now skips more 128-bit instructions

also asserts that no 128-bit int types are used, in this case
`size8u_t` and `size8s_t`.

Signed-off-by: Anton Johansson <anjo@rev.ng>

Commit:
Cast if necessary when performing tcg load ops

Needed to avoid warnings when loading 1,2, and 4 bytes from memory
into 8 byte variables.

Signed-off-by: Anton Johansson <anjo@rev.ng>

Commit:
Added checks for variable declarations

This was done by asserting that a variable exists when lookups are done
using `find_variable`. All variables are now asserted to be declared
except for the effective address `EA`. The variable `EA` is declared
when it appears in an assignment, as previously. A better way to this
might be to declare it as a 32-bit int whenever it appears in the
argument list of a function.

Signed-off-by: Anton Johansson <anjo@rev.ng>

Commit:
Rename `decl` to `arg_decl`

to better differentiate argument and variable declarations. `arg_decl`
is also easier to search for as it is more unique.

Signed-off-by: Anton Johansson <anjo@rev.ng>

Commit:
Added types to the `fADDSAT64` macro

Required as undeclared variables are no longer allowed.

Signed-off-by: Anton Johansson <anjo@rev.ng>

Commit:
Declares EA to a 32-bit signed int

when seen in the argument list. This is instead of auto
declaring it at first assignment.

Signed-off-by: Anton Johansson <anjo@rev.ng>

aleclearmind commented on target/hexagon/idef-parser/idef-parser.lex:
```diff
@@ -629,13 +633,36 @@ SIGN_ID                  s|u
+                           } else {
+                               yylval->rvalue.signedness = SIGNED;
+                           }
+                           return TYPE_SIZE_T; }
+"signed"                 { return TYPE_SIGNED; }
+"unsigned"               { return TYPE_UNSIGNED; }
+"long"                   { return TYPE_LONG; }
+"int"                    { return TYPE_INT; }
```
Are you sure you don't want a single token for all of these guys?

aleclearmind commented on target/hexagon/idef-parser/idef-parser.lex:
```diff
@@ -629,13 +633,36 @@ SIGN_ID                  s|u
+"size"[1248][us]"_t"     { /* Handles "size_t" variants of int types */
+                           const unsigned int bits_per_byte = 8;
+                           const unsigned int bytes = yytext[4] - '0';
+                           yylval->rvalue.bit_width = bits_per_byte*bytes;
```
```suggestion
                           yylval->rvalue.bit_width = bits_per_byte * bytes;
```
Use `script/checkpatch.pl` to validate your patches.

AntonJohansson replied:
Fixed formatting issues. Thanks!

aleclearmind commented on target/hexagon/idef-parser/idef-parser.y:
```diff
@@ -143,6 +145,101 @@ var : VAR
     ;
 
+/*
+ * Here the integer types are defined from valid combinations of
+ * "signed", "unsigned", "int", and "long" tokens.  The "signed"
+ * and "unsigned" tokens are here assumed to always be placed
+ * first in the type declaration, which is not the case in
+ * normal C. Similarly, "int" is assumed to be placed in the type.
+ */
+type_int : TYPE_INT
+         | TYPE_SIGNED
+         | TYPE_SIGNED TYPE_INT;
+type_uint : TYPE_UNSIGNED
+          | TYPE_UNSIGNED TYPE_INT;
+type_long : TYPE_LONG
+          | TYPE_LONG TYPE_INT
+          | TYPE_SIGNED TYPE_LONG
+          | TYPE_SIGNED TYPE_LONG TYPE_INT;
+type_ulong : TYPE_UNSIGNED TYPE_LONG
+           | TYPE_UNSIGNED TYPE_LONG TYPE_INT;
+type_longlong : TYPE_LONG TYPE_LONG
+              | TYPE_LONG TYPE_LONG TYPE_INT
+              | TYPE_SIGNED TYPE_LONG TYPE_LONG
+              | TYPE_SIGNED TYPE_LONG TYPE_LONG TYPE_INT;
+type_ulonglong : TYPE_UNSIGNED TYPE_LONG TYPE_LONG
+               | TYPE_UNSIGNED TYPE_LONG TYPE_LONG TYPE_INT;
```
It seems to me that all of these could be avoided if we were to anticipate some of this logic in the parser.

AntonJohansson replied:
Hmm I'm not sure I understand what you mean here. Do you mean in the lexer by for instance having a general `TYPE_INT` token and then lexing for example
```c
"unsigned long long"    { yylval->rvalue.signedness = UNSIGNED;
                          yylval->rvalue.bit_width = 64;
                          return TYPE_INT; }
```
Implementing the int types like this would also work, I just wanted to avoid hard coding the spaces. I believe the only int types that appear in the input are `int`, `unsigned int`, `unsigned long long`, and the`size_t`'s. All of these could be covered this way by a single `TYPE_INT` token so that would be nice.

 Or did you have something else in mind? 

Lastly, should we support only the int types in the input file, or should we include other combinations such as "signed long" and so on to be more future proof?

aleclearmind replied:
Mh, I see what you mean with the spaces things.
It's a bit verbose, but it makes sense.

aleclearmind commented on target/hexagon/idef-parser/macros.inc:
```diff
@@ -87,11 +87,11 @@
-    __sum = __a + __b;                                                  \
-    __xor = __a ^ __b;                                                  \
-    __mask = fCAST8s(0x8000000000000000ULL);                            \
+    size8u_t __a = fCAST8u(A);                                          \
+    size8u_t __b = fCAST8u(B);                                          \
+    size8u_t __sum = __a + __b;                                         \
+    size8u_t __xor = __a ^ __b;                                         \
+    size8s_t __mask = fCAST8s(0x8000000000000000ULL);                   \
```
Does this change come from the official (original) macro?

AntonJohansson replied:
Pretty much, the only difference is the type of `__mask` which is `const size8u_t` in the original as the `fCAST8s` is not present there. The macros are implemented slightly differently between ours and the original.

I guess all these could be made constant in our case.

aleclearmind replied:
Let's deviate as little as possible from official implementation of the macros, unless there's a reason.
Changing stuff is a liability.
If we can handle the original one, make it identical.

AntonJohansson replied:
Good idea, changed to the "official" version of the macro!

aleclearmind commented on target/hexagon/idef-parser/parser-helpers.c:
```diff
@@ -427,6 +427,16 @@ int find_variable(Context *c, YYLTYPE *locp, HexValue *varid)
     return -1;
 }
 
+static unsigned find_variable_and_assert_declared(Context *c,
```
You sure this functionality should not be embedded in `find_variable` entirely?
Seems like a reasonable safety check we always want.

AntonJohansson replied:
`find_variable` is only used in a single place (actually two but I'm removing one of them) and that is when declaring a new variable. It's needed to check that the variable isn't already declared, so it's good to have a version that doesn't assert. Unless we have some other way to check if a variable is already declared, I think it's best to keep both versions.

aleclearmind replied:
I see, I think I'm just not super happy with the naming then.
If you have two versions of a function, one that can fail and one that cannot, usually you go for `try_get_something` and `get_something` (which calls `try_get_something` and `abort` in case of failure).

AntonJohansson replied:
Alright, I'll move it so `find_variable` asserts and `try_find_variable` can fail

aleclearmind commented on target/hexagon/idef-parser/parser-helpers.c:
```diff
@@ -1873,27 +1899,39 @@ void gen_load(Context *c, YYLTYPE *locp, HexValue *size,
     OUT(c, locp, "tcg_gen_qemu_ld", size_suffix, sign_suffix);
-    OUT(c, locp, "(", dst, ", ", ea, ", 0);\n");
+    OUT(c, locp, "(");
+    if (dst->bit_width > size->imm.value*8) {
```
```suggestion
    if (dst->bit_width > size->imm.value * 8) {
```

AntonJohansson replied:
Fixed formatting issues. Thanks!

Commit:
Fixed some formatting

Signed-off-by: Anton Johansson <anjo@rev.ng>

Commit:
Moved assertion of var decl to `gen_varid_allocate`

`find_variable` is then only used in `gen_varid_allocate` and could be
made static to `parser-helpers.c`.

Signed-off-by: Anton Johansson <anjo@rev.ng>

Commit:
Renamed `find_variable*` functions

Signed-off-by: Anton Johansson <anjo@rev.ng>

Commit:
Changed string -> int conversion function in lexer

Now uses `strtoull` for all literals since the sign of the int is
handled explicitly in the parser using the `-` token. By reading in all
ints as unsigned 64-bit we avoid accidental casting bugs.

Also now uses a base of `0` in `strtoull` instead of explicitly setting base `10` and
`16`. Base `0` encompass these by using base 16 when the string starts
with `0x`, base 8 if it starts with `0`, and base 10 otherwise.

Signed-off-by: Anton Johansson <anjo@rev.ng>

Commit:
Changed to "offical" `fADDSAT64` macro

Required lexing `const`, however no token is emitted. We now use
exactly the macro as specified in `imported/macros.def`.

Signed-off-by: Anton Johansson <anjo@rev.ng>

Commit:
Changed `gen_rvalue_var` to use `find_variable`

instead of reimplementing it in the function.

Signed-off-by: Anton Johansson <anjo@rev.ng>

Commit:
`yyassert` now calls `abort()`

instead of only reporting the error, and then continuing.

Signed-off-by: Anton Johansson <anjo@rev.ng>
